File: flask/emptybox.py
import urllib3
import boto3
import click
import uuid
import json
import cgi
import io
import os
from flask import Flask, redirect, request
from flask.cli import with_appcontext

# Settings
app = Flask(__name__)
MAIN_SITE = os.getenv('EMPTYBOX_SITE', 'http://localhost:3000/')
S3_HOST = os.getenv('EMPTYBOX_S3_HOST')
S3_BUCKET = os.getenv('EMPTYBOX_S3_BUCKET')
S3_ACCESS_KEY = os.getenv('EMPTYBOX_S3_ACCESS_KEY')
S3_SECRET_KEY = os.getenv('EMPTYBOX_S3_SECRET_KEY')

s3 = boto3.client('s3', endpoint_url=f'http://{S3_HOST}/',
                  aws_access_key_id=S3_ACCESS_KEY,
                  aws_secret_access_key=S3_SECRET_KEY)


# Utils

def gen_key(filename):
    '''
    Generate UUID4 based S3 key. The generated key also account for file
    extension for convenience.
    >>> get_key('anime.png')
    'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.png'
    >>> get_key('citra.JPG')
    'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.JPG'
    >>> get_key('.zshrc')
    'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx.zshrc'
    >>> get_key('Dockerfile')
    'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
    >>> get_key('')
    'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
    >>> get_key(None)
    'xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx'
    '''
    if filename:
        root, ext = os.path.splitext(filename)

        if ext:
            return str(uuid.uuid4()) + ext

        elif root[0] == '.':
            return str(uuid.uuid4()) + root

    return str(uuid.uuid4())


# Handlers

@app.route('/')
def handle_default():
    return redirect(MAIN_SITE)


@app.route('/upload', methods=['POST'])
def upload():
    try:
        if request.form['type'] == 'file':
            file = request.files['file']
            body = file.stream
            filename = file.filename

        elif request.form['type'] == 'url':
            url = request.form['url']
            pool = urllib3.connection_from_url(url)
            response = pool.request('GET', url)
            body = io.BytesIO(response.data)

            filename = None

            if 'Content-Disposition' in response.headers:
                content_disposition = response.headers['Content-Disposition']
                value, params = cgi.parse_header(content_disposition)

                filename = params.get('filename')

        else:
            raise KeyError

    except KeyError as err:
        app.logger.debug(err)
        err = 'Invalid request'

        data = {'msg': err}
        return data, 400

    key = gen_key(filename)

    response = s3.put_object(Bucket=S3_BUCKET,
                             Key=key,
                             Body=body)

    if response['ResponseMetadata']['HTTPStatusCode'] != 200:
        app.logger.info(response)
        return 'Error', 400

    return {'msg': 'Saved', 'filename': key}


@app.route('/stats')
def stats():
    response = s3.list_objects_v2(Bucket=S3_BUCKET)

    return {'fileCount': response['KeyCount']}


# Scripts

@click.command('init_bucket')
@with_appcontext
def init_bucket():
    '''
    Initialize the bucket used for the application.
    '''
    s3.create_bucket(Bucket=S3_BUCKET)

    bucket_policy = {
        'Version': '2012-10-17',
        'Statement': [{
            'Effect': 'Allow',
            'Principal': '*',
            'Action': ['s3:GetObject'],
            'Resource': f'arn:aws:s3:::{S3_BUCKET}/*'
        }]
    }

    bucket_policy = json.dumps(bucket_policy)
    s3.put_bucket_policy(Bucket=S3_BUCKET, Policy=bucket_policy)

    click.echo('Successfully initialized bucket')


app.cli.add_command(init_bucket)

File: go/internal/emptybox/emptybox.go
package emptybox

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"mime"
	"net/http"
	"os"
	"path/filepath"

	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/credentials"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/s3"
	"github.com/google/uuid"
)

const (
	defaultListenAddress = "127.0.0.1:8000"
)

var (
	mainSiteAddress = getEnvOrDefault("EMPTYBOX_SITE", "http://localhost:3000")
	s3Host          = os.Getenv("EMPTYBOX_S3_HOST")
	s3Bucket        = os.Getenv("EMPTYBOX_S3_BUCKET")
	s3AccessKey     = os.Getenv("EMPTYBOX_S3_ACCESS_KEY")
	s3SecretKey     = os.Getenv("EMPTYBOX_S3_SECRET_KEY")
	s3client        = *newS3Client()

	errKey = errors.New("KeyError")
)

func index(w http.ResponseWriter, r *http.Request) {
	http.Redirect(w, r, mainSiteAddress, http.StatusPermanentRedirect)
}

func upload(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodPost {
		jsonEnc := json.NewEncoder(w)

		uploadType := r.FormValue("type")
		var uploadFile io.ReadCloser
		var fileName string

		defer func() {
			if r := recover(); r != nil {
				if err, ok := r.(error); ok && err == errKey {
					log.Printf("%v", err)
					errData := map[string]string{"msg": "Invalid Request"}

					w.WriteHeader(http.StatusBadRequest)
					jsonEnc.Encode(errData)
				}
			}
		}()

		if uploadType == "file" {
			mulFile, mulFileHeader, err := r.FormFile("file")
			if err != nil {
				panic(errKey)
			}

			uploadFile = mulFile
			fileName = mulFileHeader.Filename

		} else if uploadType == "url" {
			url := r.FormValue("url")
			response, err := http.Get(url)
			if err != nil {
				panic(errKey)
			}
			contentDisposition := response.Header.Get("Content-Disposition")
			_, params, err := mime.ParseMediaType(contentDisposition)
			if err != nil {
				panic(err)
			}

			uploadFile = response.Body
			fileName, _ = params["filename"]

		} else {
			panic(errKey)
		}

		content, err := ioutil.ReadAll(uploadFile)
		if err != nil {
			panic(err)
		}

		body := bytes.NewReader(content)
		key := genKey(fileName)
		input := s3.PutObjectInput{Bucket: &s3Bucket, Key: &key, Body: body}
		s3client.PutObject(&input)

		saveData := map[string]string{
			"msg":      "Saved",
			"filename": key,
		}

		jsonEnc.Encode(saveData)

	} else {
		w.WriteHeader(http.StatusMethodNotAllowed)
	}
}

func stats(w http.ResponseWriter, r *http.Request) {
	input := s3.ListObjectsV2Input{Bucket: &s3Bucket}
	ls, err := s3client.ListObjectsV2(&input)
	if err != nil {
		fmt.Println(err.Error())
		panic(err)
	}

	statsData := map[string]*int64{"fileCount": ls.KeyCount}

	jsonEnc := json.NewEncoder(w)
	jsonEnc.Encode(statsData)
}

func ServeHTTP(bindAddress string) {
	server := http.NewServeMux()
	server.HandleFunc("/", index)
	server.HandleFunc("/upload/", upload)
	server.HandleFunc("/stats/", stats)

	log.Print("========================= emptybox =========================")
	log.Printf("Server starting at %s\n", bindAddress)
	http.ListenAndServe(bindAddress, server)
}

// genKey generate UUID4 based S3 key. The generated key also account for file
// extension for convenience.
func genKey(key string) string {
	u, err := uuid.NewRandom()

	if err != nil {
		fmt.Println(err.Error())
		panic(err)
	}

	if key != "" {
		base := filepath.Base(key)
		ext := filepath.Ext(key)

		if ext != "" {
			return u.String() + ext
		}

		if base[0] == '.' {
			return u.String() + base
		}
	}

	return u.String()
}

func getEnvOrDefault(key, fallback string) string {
	if value, ok := os.LookupEnv(key); ok {
		return value
	}

	return fallback
}

func setAWSCredentialsEnv() {
	os.Setenv("AWS_ACCESS_KEY_ID", s3AccessKey)
	os.Setenv("AWS_SECRET_ACCESS_KEY", s3SecretKey)
}

func newS3Client() *s3.S3 {
	// Configure to use MinIO Server
	s3Config := &aws.Config{
		Credentials:      credentials.NewStaticCredentials(s3AccessKey, s3SecretKey, ""),
		Endpoint:         aws.String(s3Host),
		Region:           aws.String("us-east-1"),
		DisableSSL:       aws.Bool(true),
		S3ForcePathStyle: aws.Bool(true),
	}

	newSession := session.New(s3Config)
	return s3.New(newSession)
}

func InitBucket() {
	_, err := s3client.CreateBucket(&s3.CreateBucketInput{
		Bucket: &s3Bucket,
	})

	if err != nil {
		fmt.Println(err.Error())
		panic(err)
	}

	readOnlyAnonUserPolicy := map[string]interface{}{
		"Version": "2012-10-17",
		"Statement": []map[string]interface{}{
			{
				"Sid":       "AddPerm",
				"Effect":    "Allow",
				"Principal": "*",
				"Action": []string{
					"s3:GetObject",
				},
				"Resource": []string{
					fmt.Sprintf("arn:aws:s3:::%s/*", s3Bucket),
				},
			},
		},
	}

	policy, err := json.Marshal(readOnlyAnonUserPolicy)
	if err != nil {
		panic(err)
	}

	_, err = s3client.PutBucketPolicy(&s3.PutBucketPolicyInput{
		Bucket: &s3Bucket,
		Policy: aws.String(string(policy)),
	})

	if err != nil {
		fmt.Println(err.Error())
		panic(err)
	}
}
